letsencrypt/tests/auth_handler_test.py

```diff
@@ -148,6 +152,66 @@ def test_perform_failure(self):
         self.assertRaises(
             errors.AuthorizationError, self.handler.get_authorizations, ["0"])
 
+    @mock.patch("acme.challenges.HTTP01Response.simple_verify")
+    @mock.patch("acme.challenges.TLSSNI01Response.simple_verify")
+    @mock.patch("letsencrypt.auth_handler.logger")
+    def test_simple_verify(self, mock_logger, mock_tlssni01_verify, mock_http01_verify):
```

Comment:
Not everything can be tested in a single run. That's exactly why I think you need separate tests.

The bogus code I wrote can't really exist by mistake, but it is an extreme to show everything that isn't tested currently. I wasn't trying to fool the test inconspicuously, I was trying to fool the test easily. A better test is probably harder to fool (and thus also harder to fool unintentionally). I'm also not saying every possibility has to be tested, but that it is desirable to better safeguard the intended behavior here.

For example you could create the following tests:
 - **test_simple_verify_one_success_http01** Call the code with 1 mock_http01 that returns True. Assert that it does not log the warning.
 - **test_simple_verify_one_success_tlssni01** Call the code with 1 mock_tlssni01 that returns True. Assert that it does not log the warning.
 - **test_simple_verify_one_failure** Call the code with 1 mock that returns False. Assert that it logs the warning.
 - **test_simple_verify_one_unimplemented** Call the code with the dummy. Assert that it logs the debug message.
 - **test_simple_verify_many** Call the code with multiple challenges. Assert that the correct number of calls are made.

That's a lot of tests but this is only an example and more or less tests could also be fine. The above tests could be short and simple and hopefully easy to read. If I would know the Lets Encrypt code a bit better I would be able te better predict what @schoen or @kuba is going to expect :-)